iceoryx2-bb/posix/src/periodic_timer.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! # Example
+//!
+//! ```no_run
+//! use iceoryx2_bb_periodic_timer::*;
+//!
+//! let periodic_timer = PeriodicTimerBuilder::new().create().unwrap();
+//!
+//! // the timer waits on the following time points
+//! // 4 5 8 9 10 12 15 16 18
+//!
+//! periodic_timer.add(Duration::from_secs(4));
+//! periodic_timer.add(Duration::from_secs(5));
+//! periodic_timer.add(Duration::from_secs(9));
+//!
+//! std::thread::sleep(periodic_timer.next_iteration().unwrap());
+//!
+//! ```
+
+use std::{cell::RefCell, sync::atomic::Ordering, time::Duration};
+
+use iceoryx2_bb_log::fail;
+use iceoryx2_pal_concurrency_sync::iox_atomic::IoxAtomicU64;
+
+use crate::{
+    clock::ClockType,
+    clock::{Time, TimeError},
+};
+
+/// Represents an index to identify an added timer with [`PeriodicTimer::add()`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PeriodicTimerIndex(u64);
+
+/// Represents the RAII guard of [`PeriodicTimer`] and is returned by [`PeriodicTimer::add()`].
+/// As soon as it goes out of scope it removes the attached cyclic timeout from [`PeriodicTimer`].
+pub struct PeriodicTimerGuard<'periodic_timer> {
```

Comment:
Can this be combined with `Attachment` ? The caller could get an `Attachment` which enables RAII clean-up and also manual removal from the timer (instead of an index). I like `ScheduledInterval` better for the name.

Reply:
No, since the `Attachment` is an internal class that is used by the `PeriodicTimer` to store internal management stuff about the thing that was attached.

Manual removal wouldn't be required, the user can always use `drop(guard);`. I could also make the `remove` function public but I do not see the gain of it.

Reply:
Fine as internal. Maybe revisit later.